2.0.0 (#31)
kirkone/KK.AspNetCore.EasyAuthAuthentication: README.md

        options.DefaultAuthenticateScheme = EasyAuthAuthenticationDefaults.AuthenticationScheme;
        options.DefaultChallengeScheme = EasyAuthAuthenticationDefaults.AuthenticationScheme;
    }
).AddEasyAuth(this.Configuration);
```

The next you must add the easy auth middleware to the ASP.NET pipeline. This can be enabled by the folling section in your `Configure`method in the `Startup`:

```csharp
app.UseAuthentication();
```

> Warning: make sure you add the `UseAuthentication` in the right order for your use case.

The last thing is to add the following section in the `appsettings.json` to enabled the basic providers:

```json
"easyAuthOptions": {

>   "providerOptions": [

Reviewer: Shouldn't be the line 
"AuthEndpoint": ".auth/me",
here as well?

Author: No @paule96 changed this in the last session. maybe the is a little catch with the implementation

Reviewer: Hm, actually I get the web app not started, because it wants an .auth folder in wwwroot, which isn't there. This folder is only used for local debugging, when i have the correct understanding. In azure it is mapped by the surrounding web app, isn't it?

Reviewer 2: There is no check if the folder is availed. Please look in to your looks. Maybe share the information with us. There are some new checks. 

We first test if you has configured you web app in the right way. If not it dies.

Then if a request is send to a protected endpoint and can't authorize it end the request, but the web app won't die

Reviewer: When configuring easy auth like this: ArgumentNullException: Value cannot be null.<br />\r\nParameter name: The 'AuthEndpoint' in the configuraiton for easy auth can't be empty! please add the setting to your configuration providers.

Reviewer: It is no problem to skip AuthEndpoint in Config, when using EasyAuthWithHeaderService. That's what I could see here with my WebApps in MCD.

Reviewer 2: After a call with you, we maybe find the bug. Please share the result of your testing here so we can close this conversation.

Keep rocking. 🚀🧙‍♂️

Reviewer: Now it looks good, authentication in my web apps works so far i can see. Thanks, Paule96! I have not tested, what happens if I come with an valid, but expired token to an app which uses EasyAuthForAuthorizationTokenService, but for normal authentification it works great.

```json
"easyAuthOptions": {
    "AuthEndpoint": ".auth/me",
    "providerOptions": [
      {
        "ProviderName": "EasyAuthForAuthorizationTokenService",
        "Enabled": true,
        "NameClaimType": "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/spn", // optional
        "RoleClaimType": "roles" // optional
      },
      {
        "ProviderName": "EasyAuthWithHeaderService",
        "Enabled": true,
        "NameClaimType": "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress", //optional
        "RoleClaimType": "http://schemas.microsoft.com/ws/2008/06/identity/claims/role" //optional
      }
    ]
  }
```

Here are some notes to the JSON above:

- each provider is disabled by default so you must enabled it
- you can create own providers but there must implement `IEasyAuthAuthentificationService`. But you must also activate them here. (Don't put them in the DI. This package will do this by it's own.)
- The `ProviderName` is the class name of the provider. that must be unique in your application.

> > A list of all providers can be found in the headline `Auth Provider`

Reviewer: Add a description how to use the EasyAuthWithAuthMeService for local debugging.

### Configure options via configuration (recommended)

You can use the default behavior of asp.net core to configure EasyAuth. You must only change in your `Startup.cs` the `.AddEasyAuth()` to `.AddEasyAuth(this.Configuration)`.

> To get the property `this.Configuration` in your `Startup.cs` you must add `IConfiguration configuration` to your constructor parameters and create a property.

To configure you providers you simple add the following to your appsettings.json. (or to your environment variables, or other [configuration sources](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/).)

```json
"easyAuthOptions": {
    "AuthEndpoint": ".auth/me",
    "providerOptions": [
      {
        "ProviderName": "EasyAuthForAuthorizationTokenService",
        "Enabled": true,
        "NameClaimType": "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/spn", // optional
        "RoleClaimType": "roles" // optional
      },
      {
        "ProviderName": "EasyAuthWithHeaderService",
        "Enabled": true,
        "NameClaimType": "http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress", //optional

>         "RoleClaimType": "http://schemas.microsoft.com/ws/2008/06/identity/claims/role" //optional

Reviewer: Is it by purpose, that the ClaimsPrincipal will have "new" role claims from type ClaimsType.Role after successful authentification? Even if I submit "RoleClaimType": "roles" here? 
Or to say it in another way: 
Is this setting only means the role claim type, which comes in by token(, cookie, header, ...) and has no effect on the output claims of the ClaimsPrincipal after authentification?

Reviewer 2: The Output is always Role. See this [line](https://github.com/kirkone/KK.AspNetCore.EasyAuthAuthentication/blob/0ad56bfb858933276abeeed27784d522e0fef0b6/src/KK.AspNetCore.EasyAuthAuthentication/AuthenticationTicketBuilder.cs#L45). You can only define what is the input. So you can map groups for example to roles. But in your application it will always the role claim. Should we change this? Or is this Behavior okay? Maybe you can open an issue for that?

I will now add the documentation for that.

Reviewer 2: see the commit https://github.com/kirkone/KK.AspNetCore.EasyAuthAuthentication/pull/31/commits/a786c7e1e58dfb384f2bad75ae8d99115713c3ee for the documentation changes

> **Info**: Make sure you added static file handling to your pipeline by adding `app.UseStaticFiles();` to your `public void Configure` method in the `Startup.cs`, e.g. just after `app.UseHttpsRedirection();` entry. Otherwise the static file can not be found at runtime.

> **Info**: Using a wwwroot sub-folder name that starts with `'.'`, like the suggested `.auth` folder name, is useful for content relevant only for localhost debugging as these are treated as hidden folders and are not included in publish output.

## Auth Provider

There are some predefined providers in this package. If you need your own or want contribute to our existing providers you must implement the `IEasyAuthAuthentificationService`.

### `EasyAuthWithAuthMeService` (always on)

> This is a slightly special provider. This provider cannot be configured and does not implement the 'IEasyAuthAuthentificationService'. This provider is for development. A developer can create a JSON with the content of the `/.auth/me` endpoint of an EasyAuth Azure Web App. So you don't need a connection to the internet or azure for development and just use your local things.

Reviewer: This isn't true. You must configure the EasyAuthWithAuthMeService for local debugging. Otherwise it won't work.

> **Info**: Make sure you added static file handling to your pipeline by adding `app.UseStaticFiles();` to your `public void Configure` method in the `Startup.cs`, e.g. just after `app.UseHttpsRedirection();` entry. Otherwise the static file can not be found at runtime.

> **Info**: Using a wwwroot sub-folder name that starts with `'.'`, like the suggested `.auth` folder name, is useful for content relevant only for localhost debugging as these are treated as hidden folders and are not included in publish output.

## Auth Provider

There are some predefined providers in this package. If you need your own or want contribute to our existing providers you must implement the `IEasyAuthAuthentificationService`.

### `EasyAuthWithAuthMeService` (always on)

> This is a little bit special provider. That provider can't be configured and it isn't implementing `IEasyAuthAuthentificationService`. This provider is for the development case, so a developer can create a JSON with the content of the `/.auth/me` endpoint of an EasyAuth Azure Web App. So you don't need a internet connection or azure for development and can use only local things.

Reviewer 3: ```suggestion
This is a slightly special provider. This provider cannot be configured and does not implement the 'IEasyAuthAuthentificationService'. This provider is for development. A developer can create a JSON with the content of the `/.auth/me` endpoint of an EasyAuth Azure Web App. So you don't need a connection to the internet or azure for development and just use your local things.
```

### `EasyAuthForAuthorizationTokenService`

> This provider is for the case you have a Azure Web App that is not only be used by humans. so maybe you want access your app with an Service Principal (SPN). If you enabled this provider you can access your app with Azure Ad Service Principals (SPN).

Reviewer 3: ```suggestion
 Use this provider if you have a Azure Web App, that is not only be used by humans. For instance if you want to access your app with a Service Principal (SPN). 
```

> To create an Service Principal (SPN) that can have access to your EasyAuth protected Application you must change the app manifest for you application in your Azure AD. Thanks to [Suzuko123](https://github.com/Suzuko123) for the following sample:

Reviewer 3: ```suggestion
To create an Service Principal (SPN), which get access to your EasyAuth protected Application. You have to change the app manifest for you application in your Azure AD. Thanks to [Suzuko123](https://github.com/Suzuko123) for the following sample:
```

```json
"appRoles": [
	{
	    "allowedMemberTypes": [
			"Application"
		],
	    "description": "allow a call as system admin.",
		"displayName": "SystemAdmin",
		"id": "dd6d2784-5fa1-4c97-9f9b-8376a85b4163",
		"isEnabled": true,
		"lang": null,
		"origin": "Application",
		"value": "SystemAdmin"
	}
]
```

This will allow a spn to get the role `SystemAdmin` in your protected application. The default `User.Identity.Name` of an SPN is the SPN Guid.

### `EasyAuthWithHeaderService`

> This is the most common auth provider. This let you use Azure Active Directory Users in your easy auth application.

Reviewer 3: ```suggestion
This is the most common auth provider. You can use it to work with Azure Active Directory Users in your easy auth application.
```
